lib/kubernetes-deploy/container_logs.rb

```diff
@@ -0,0 +1,81 @@
+# frozen_string_literal: true
+module KubernetesDeploy
+  class ContainerLogs
+    attr_reader :lines, :container_name
+
+    DEFAULT_LINE_LIMIT = 250
+
+    def initialize(parent_id:, container_name:, logger:)
+      @parent_id = parent_id
+      @container_name = container_name
+      @logger = logger
+      @lines = []
+      @last_printed_index = -1
+    end
+
+    def sync(kubectl)
+      new_logs = fetch_latest(kubectl)
+      return unless new_logs.present?
+      @lines += deduplicate(new_logs)
+    end
+
+    def empty?
+      lines.empty?
+    end
+
+    def print_latest(prefix: false)
+      prefix_str = "[#{container_name}]  " if prefix
+      start_at = @last_printed_index + 1
+
+      lines[start_at..-1].each do |msg|
+        @logger.info "#{prefix_str}#{msg}"
+      end
+
+      @last_printed_index = lines.length - 1
+    end
+
+    def print_all
+      lines.each { |line| @logger.info("\t#{line}") }
+    end
+
+    private
+
+    def fetch_latest(kubectl)
+      cmd = ["logs", @parent_id, "--container=#{container_name}", "--timestamps"]
+      cmd << if @last_timestamp.present?
+        "--since-time=#{rfc3339_timestamp(@last_timestamp)}"
+      else
+        "--tail=#{DEFAULT_LINE_LIMIT}"
+      end
+      out, _err, _st = kubectl.run(*cmd, log_failure: false)
+      out.split("\n")
+    end
+
+    def rfc3339_timestamp(time)
+      time.strftime("%FT%T.%N%:z")
```

Comment:
Looks like Ruby has a method for this. Should we use that?

https://ruby-doc.org/stdlib-2.3.1/libdoc/date/rdoc/DateTime.html#method-i-rfc3339

Reply:
The original code did, but [Policial didn't like it](https://policial.shopify.io/Shopify/kubernetes-deploy/builds/695493). TBH I got annoyed trying to figure out whether the justification for that blanket prohibition actually applies here, and made this method to avoid going through DateTime in any case. Looking again, the [PR](https://github.com/rubocop-hq/rubocop/issues/1952) description doesn't explain anything, but further down there's a discussion about DateTime not handling some timezones properly. I don't think that does matter here since the timestamps I'm getting back (locally and in prod) are all formatted like `2018-10-08T02:04:02.851080236Z`.